### PR TITLE
Fix unnecessary simulcast cropping, vastly simplify simulcast uplink policy

### PR DESCRIPTION
**Pull request by hensmi-amazon**
**Issue #:** https://github.com/aws/amazon-chime-sdk-js/issues/2844, but after I fixed that by removing code, it somehow reduced coverage in all these corners related to our 'pausedByWebRTC' logic, which at this point is no longer needed. So I just nuked it all.

**Description of changes:** So back in the dark ages, we had this 'issue' where when simulcast was enabled, Chromium didn't really have a mechanism for indicating that layers would be paused if there wasn't enough bandwidth for them. So what we did was stare at `simulcast.cc` in libwebrtc and try to guess in JS when the layers would be dropped, and then use this information to resubscribe without the dropped stream being published. This was clunky but it did the job.

However when https://github.com/aws/amazon-chime-sdk-js/commit/2f8c523fd708837be32ef86dbd1ba407099d4b2e was released (the second time :) ) it gave us a mechanism to indicate to backend when a stream was paused. So all the aforemention logic is now completely unnecessary. Note that I'm pretty sure we don't need to resubscribe on layers switches anymore but I will do that another day.

**Testing:**

*Can these tested using a demo application? Please provide reproducible step-by-step instructions.*
* Enable simulcast and vary senders from 1-8.
* On the receive side, switch between layers by enabling the priority based downlink policy.
* Should not be able to switch layers when call is 1-1 since we will disable simulcast
* Test applying network constraints to uplink side and ensure that there are no long freezes

**Checklist:**

1. Have you successfully run `npm run build:release` locally?
y

2. Do you add, modify, or delete public API definitions? If yes, has that been reviewed and approved?
n

3. Do you change the wire protocol, e.g. the request method? If yes, has that been reviewed and approved?
n

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: demos/browser/app/meetingV2/meetingV2.ts [hunk 1]
```diff
@@ -3968,7 +3968,7 @@ export class DemoMeetingApp
 
     const chosenVideoSendCodec = (document.getElementById('videoCodecSelect') as HTMLSelectElement).value;
     this.videoCodecPreferences = getCodecPreferences(chosenVideoSendCodec);
-    if (!['av1Main', 'vp9Profile0'].includes(chosenVideoSendCodec)) {
+    if (['av1Main', 'vp9Profile0'].includes(chosenVideoSendCodec)) {
```
Comment (hensmi-amazon) [inline demos/browser/app/meetingV2/meetingV2.ts:3971]: Unrelated fix

File: src/task/MonitorTask.ts [hunk 2]
```diff
@@ -179,7 +179,6 @@ export default class MonitorTask
           `trigger resubscribe for up=${resubscribeForUplink}; videosToReceive=[${this.context.videosToReceive.array()}]`
         );
         this.context.videoUplinkBandwidthPolicy.chooseEncodingParameters();
-        this.context.videoUplinkBandwidthPolicy.chooseMediaTrackConstraints();
```
Comment (hensmi-amazon) [inline src/task/MonitorTask.ts:182]: As seen below, this is now a no op in every implementation.

File: src/task/ReceiveVideoInputTask.ts [hunk 3]
```diff
@@ -120,24 +120,13 @@ export default class ReceiveVideoInputTask extends BaseTask {
       );
       const trackSettings = videoTracks[0].getSettings();
 
-      if (this.context.enableSimulcast) {
-        // For video, we currently let uplink policy to specify constraints for simulcast. This logic should be removed in the future.
```
Comment (hensmi-amazon) [inline src/task/ReceiveVideoInputTask.ts:124]: I checked with Shi and he agreed this divergence wasn't necessary. I think we were just playing it safe, initially.

File: src/videouplinkbandwidthpolicy/DefaultSimulcastUplinkPolicy.ts [hunk 4]
```diff
@@ -82,160 +62,84 @@ export default class DefaultSimulcastUplinkPolicy implements SimulcastUplinkPoli
...
-    } else if (this.currentActiveStreams === ActiveStreams.kHiAndLow) {
-      hysteresisIncrease = this.enableFhdVideo ? 4000 : 2400;
-      hysteresisDecrease = this.hiDisabledRate;
-    } else if (this.currentActiveStreams === ActiveStreams.kMidAndLow) {
-      hysteresisIncrease = this.enableFhdVideo ? 1600 : 1000;
-      hysteresisDecrease = this.midDisabledRate;
+    if (this.shouldDisableSimulcast) {
+      // See comment above `shouldDisableSimulcast` for usage.
+      //
+      // The value of `newActiveStreams` is somewhat irrelevant since in one to one calls
```
Comment (hensmi-amazon) [inline src/videouplinkbandwidthpolicy/DefaultSimulcastUplinkPolicy.ts:78]: I think i mitigated this issue with some recent backend changes. I'll remove this mega comment another time though.

File: src/videouplinkbandwidthpolicy/DefaultSimulcastUplinkPolicy.ts [hunk 5]
```diff
@@ -82,160 +62,84 @@ export default class DefaultSimulcastUplinkPolicy implements SimulcastUplinkPoli
...
     }
     return this.newQualityMap;
   }
 
   chooseMediaTrackConstraints(): MediaTrackConstraints {
-    // Changing MediaTrackConstraints causes a restart of video input and possible small
-    // scaling changes.  Always use 720p for now
-    const trackConstraint: MediaTrackConstraints = {
-      width: { ideal: this.enableFhdVideo ? 1920 : 1280 },
-      height: { ideal: this.enableFhdVideo ? 1080 : 768 },
```
Comment (hensmi-amazon) [inline src/videouplinkbandwidthpolicy/DefaultSimulcastUplinkPolicy.ts:230]: This 768 value was the original issue. Leads to cropping. But 720 works just fine.

File: demos/browser/app/meetingV2/meetingV2.ts (hunk 1 above)
Comment (ltrung) [reply demos/browser/app/meetingV2/meetingV2.ts:3971]: Oops :D

File: src/videouplinkbandwidthpolicy/VideoUplinkBandwidthPolicy.ts [hunk 6]
```diff
@@ -61,7 +61,7 @@ export default interface VideoUplinkBandwidthPolicy {
   updateConnectionMetric(metrics: ConnectionMetrics): void;
 
   /**
-   * Returns the selected [[MediaTrackConstraints]] to update
+   * This function is deprecated and unused, and will be removed in a future release.
```
Comment (ltrung) [inline src/videouplinkbandwidthpolicy/VideoUplinkBandwidthPolicy.ts:64]: nit: Can we reformat as TODO? in major version, we usually just do a TODO search for where we need to remove.

Comment (hensmi-amazon) [reply src/videouplinkbandwidthpolicy/VideoUplinkBandwidthPolicy.ts:64]: I think we banned TODOs at some point lol. I've been intentionally using the word 'deprecated' on these types of functions to make them easier to find in the future.

File: src/videouplinkbandwidthpolicy/DefaultSimulcastUplinkPolicy.ts [hunk 7]
```diff
@@ -82,160 +62,84 @@ export default class DefaultSimulcastUplinkPolicy implements SimulcastUplinkPoli
...
   }
 
-  updateConnectionMetric({ uplinkKbps = 0 }: ConnectionMetrics): void {
-    if (isNaN(uplinkKbps)) {
-      return;
-    }
-
-    // Check if startup period in order to ignore estimate when video first enabled.
-    // If only audio was active then the estimate will be very low
-    if (this.startTimeMs === 0) {
```
Comment (ltrung) [inline src/videouplinkbandwidthpolicy/DefaultSimulcastUplinkPolicy.ts:92]: So glad to see this gone :))

File: src/videouplinkbandwidthpolicy/DefaultSimulcastUplinkPolicy.ts [hunk 8]
```diff
@@ -82,160 +62,84 @@ export default class DefaultSimulcastUplinkPolicy implements SimulcastUplinkPoli
...
+      //
+      // Also note that due to some uninvestigated logic in bitrate allocation, Chromium
+      // will skip the bottom layer if we try setting it to 1200 kbps instead so it will
+      // still take a while to recover (as it needs to send padding until it reaches around
+      // 1000 kbps).
+      this.newActiveStreams = ActiveStreams.kHi;
+      newBitrates[0].maxBitrateKbps = 0;
+      newBitrates[1].maxBitrateKbps = this.hiTargetBitrateKbps;
+      newBitrates[2].maxBitrateKbps = 0;
+    } else if (this.numSenders <= 4) {
```
[the remark this answers is not in the document]
Comment (ltrung) [reply src/videouplinkbandwidthpolicy/DefaultSimulcastUplinkPolicy.ts:106]: To follow up, is there a world where we could just have one simple uplink policy based on number of clients :)

File: src/videouplinkbandwidthpolicy/DefaultSimulcastUplinkPolicy.ts [hunk 9]
```diff
@@ -82,160 +62,84 @@ export default class DefaultSimulcastUplinkPolicy implements SimulcastUplinkPoli
...
+      // E.g., 320x192+ (640x384)  + 1280x768
+      this.newActiveStreams = ActiveStreams.kHiAndLow;
+      newBitrates[0].maxBitrateKbps = this.lowTargetBitrateKbps;
+      newBitrates[1].maxBitrateKbps = 0;
+      newBitrates[2].maxBitrateKbps = this.hiTargetBitrateKbps;
     } else {
-      hysteresisIncrease = this.enableFhdVideo ? 500 : 300;
-      hysteresisDecrease = 0;
+      // E.g., 320x192 + 640x384 + (1280x768)
+      this.newActiveStreams = ActiveStreams.kMidAndLow;
```
Comment (ltrung) [inline src/videouplinkbandwidthpolicy/DefaultSimulcastUplinkPolicy.ts:114]: Can we add a comment for the reasoning for bitrate here?

File: src/videouplinkbandwidthpolicy/DefaultSimulcastUplinkPolicy.ts [hunk 10]
```diff
@@ -17,23 +16,13 @@ const enum ActiveStreams {
...
-  static readonly defaultUplinkBandwidthKbps: number = 1200;
-  static readonly startupDurationMs: number = 6000;
-  static readonly holdDownDurationMs: number = 4000;
-  static readonly defaultMaxFrameRate = 15;
-  // Current rough estimates where webrtc disables streams
-  static readonly kHiDisabledRateHd = 700;
-  static readonly kMidDisabledRateHd = 240;
-  static readonly kHiDisabledRateFhd = 1200;
-  static readonly kMidDisabledRateFhd = 400;
   static readonly kHiTargetBitrateKbpsHd = 1200;
```
Comment (ltrung) [inline src/videouplinkbandwidthpolicy/DefaultSimulcastUplinkPolicy.ts:26]: So there is no difference  between fhd and hd cases?

File: src/videouplinkbandwidthpolicy/DefaultSimulcastUplinkPolicyForContentShare.ts [hunk 11]
```diff
@@ -83,19 +74,6 @@ export default class DefaultSimulcastUplinkPolicyForContentShare implements Simu
     this.defaultLowTargetBitrateKbps = enabled ? 500 : 300;
   }
 
-  private getQualityMapString(params: Map<string, RTCRtpEncodingParameters>): void {
```
Comment (hensmi-amazon) [inline src/videouplinkbandwidthpolicy/DefaultSimulcastUplinkPolicyForContentShare.ts:117]: Not too useful of a debug string since these values never change now :)

**Review by shi-su — APPROVED** [review]

**Review by ltrung — APPROVED** [review]

**Comment by BrandonMathis** [comment]
I was dealing with camera resolution issues with simulcast enabled in chromium based browsers (camera was coming back from chime looking very zoomed) and this PR fixed my problem.  Thankyou 🙏🏼

Any possible updates on a release or beta release timeline for this fix to npmjs? Installing and building modules directly from GitHub is a bit challenging in my cloud environment.

**Comment by hensmi-amazon** [comment]
@BrandonMathis you can use the mitigation in https://github.com/aws/amazon-chime-sdk-js/issues/2844
